src/appleseed-max-impl/appleseedrenderer/projectbuilder.cpp

```diff
@@ -231,15 +232,38 @@ namespace
             asf::uint32 normal_indices[3];
             if (face_smgroup == 0)
             {
-                // No smooth group for this face, use the face normal.
-                const Point3 n = normal_transform * mesh.getFaceNormal(i);
-                const asf::uint32 normal_index =
-                    static_cast<asf::uint32>(
-                        object->push_vertex_normal(
-                            asf::safe_normalize(asr::GVector3(n.x, n.y, n.z))));
-                normal_indices[0] = normal_index;
-                normal_indices[1] = normal_index;
-                normal_indices[2] = normal_index;
+                // No smooth group for this face, check if face has explicit normals.
+                bool normal_set = false;
+                MeshNormalSpec* nspec = mesh.GetSpecifiedNormals();
+                if (nspec)
+                {
+                    for (int j = 0; j < 3; ++j)
+                    {
+                        const int norm_index = nspec->GetNormalIndex(i, j);
+                        if (norm_index < 0)
```

Comment:
I feel like there's a potential bug here:

Can it be that there's a normal at j == 0, in which case `normal_set` is set to true, but no normal at j == 1, in which case we break out of the loop (but `normal_set` is set to true so we won't switch to using the face normal)?

Reply:
I think I fixed it @dictoon